### PR TITLE
add support for abstract domain sockets in OpenTelemetry.Exporter.Geneva

### LINKED ISSUE #1198: add support for abstract domain sockets in OpenTelemetry.Exporter.Geneva
# Issue with OpenTelemetry.Exporter.Geneva

List of [all OpenTelemetry NuGet
packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are
using (e.g. `OpenTelemetry 1.3.2`):

``` json
      "OpenTelemetry": {
        "type": "Transitive",
        "resolved": "1.3.1",
        "contentHash": "PWIl/juBJBFi4Adx07nltkxA2ZCwutd2A+KAvO3TOYV+xeIiN6afT5bE4W3EebqdDV0T3buq/vAlQ2ZDC9zYew==",
        "dependencies": {
          "Microsoft.Extensions.Logging": "3.1.0",
          "Microsoft.Extensions.Logging.Configuration": "3.1.0",
          "OpenTelemetry.Api": "1.3.1",
          "System.Collections.Immutable": "1.4.0",
          "System.Reflection.Emit.Lightweight": "4.7.0"
        }
      },
      "OpenTelemetry.Exporter.Geneva": {
        "type": "Transitive",
        "resolved": "1.3.0",
        "contentHash": "ZSCAt9/tLMcNW1shvArvu58+7xuwairhzdl4ARWvx9+91ItYe5QLbcreDueOHdEPC5dF2nitHFuR5wdNk2HArw==",
        "dependencies": {
          "OpenTelemetry": "[1.3.0, 2.0.0)"
        }
      },
```

Runtime version (e.g. `net462`, `net48`, `net6.0`, `net7.0` etc. You can
find this information from the `*.csproj` file):

net6.0

**Is this a feature request or a bug?**

* [x] Feature Request
* [ ] Bug

**What is the expected behavior?**

I am able to export telemetry through an `mdsd` that uses the `-u` switch to "Use abstract domain sockets instead of file system paths". Perhaps a connection string like: `"Endpoint=unix:@/var/run/default_fluent.socket"`.

**What is the actual behavior?**

Start `mdsd` with the `-u` switch:
```
/usr/sbin/mdsd -A -r $HOME/mdsd -S $HOME/eh -p 0 -f 0 -D -u
```

``` csharp
var mdsdConnectionString = "Endpoint=unix:\0/home/cataggar/mdsd_fluent.socket";
```
Errors with:
```
System.Private.CoreLib: Exception has been thrown by the target of an invocation. System.Net.Sockets: Cannot assign requested address %5C0/home/cataggar/mdsd_fluent.socket.
Value cannot be null. (Parameter 'provider')
```

``` csharp
var mdsdConnectionString = "Endpoint=unix:@/home/cataggar/mdsd_fluent.socket";
```
Errors with:
```
System.Private.CoreLib: Exception has been thrown by the target of an invocation. System.Net.Sockets: Cannot assign requested address @/home/cataggar/mdsd_fluent.socket.
Value cannot be null. (Parameter 'provider')
```

## Additional Context

`netstat -a | grep mdsd` will display the abstract domain sockets with a `@` prefix. The .NET API expects a `\0` null character. Tests for abstract domain sockets are linked from https://github.com/dotnet/runtime/issues/77931#issuecomment-1304360062.

### PR DESCRIPTION
**Pull request by cataggar**
Fixes #1198.

## Changes

This allows `@` character represent the null byte used in the connection strings for abstract domain sockets.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [x] Design discussion issue #
* [x] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Exporter.Geneva/Internal/ConnectionStringBuilder.cs [hunk 1]
```diff
@@ -137,7 +137,7 @@ public string ParseUnixDomainSocketPath()
         try
         {
             var endpoint = new Uri(this.Endpoint);
-            return endpoint.AbsolutePath;
+            return endpoint.AbsolutePath.Replace('@','\0');
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Geneva/Internal/ConnectionStringBuilder.cs:140]: Only replace when `@` is the first character?

Comment (cataggar) [reply src/OpenTelemetry.Exporter.Geneva/Internal/ConnectionStringBuilder.cs:140]: Is that required to merge this?

File: src/OpenTelemetry.Exporter.Geneva/Internal/ConnectionStringBuilder.cs [hunk 2]
```diff
@@ -132,12 +132,29 @@ public TransportProtocol Protocol
...
+    /// <summary>
+    /// Replace first charater of string if it matches with <paramref name="oldChar"/> with <paramref name="newChar"/>.
+    /// </summary>
+    /// <param name="str">String to be updated.</param>
+    /// <param name="oldChar">Old character to be replaced.</param>
+    /// <param name="newChar">New character to be replaced with.</param>
+    /// <returns>Updated string.</returns>
+    internal static string ReplaceFirstChar(string str, char oldChar, char newChar)
+    {
+        if (str.Length > 1 && str[0] == oldChar)
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Geneva/Internal/ConnectionStringBuilder.cs:144]:
```suggestion
        if (str.Length >= 1 && str[0] == oldChar)
```

**Review by reyang — APPROVED** [review]
LGTM with a changelog entry.

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,8 @@
 * TldLogExporter to export `SpanId` value in `ext_dt_spanId` field instead of
   `TraceId` value.
   ([#1184](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1184))
+* Add support for abstract domain sockets
+  ([#1198](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1198))
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md:9]:
```suggestion
  ([#1199](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1199))
```

File: src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,8 @@
 * TldLogExporter to export `SpanId` value in `ext_dt_spanId` field instead of
   `TraceId` value.
   ([#1184](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1184))
+* Add support for abstract domain sockets
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md:8]:
```suggestion
* Add support for abstract domain sockets.
```

**Comment by utpilla** [comment]
Thanks for this PR @cataggar ! I have pushed two minor changes related to CHANGELOG edits: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1199/commits/384549d8061bf6d74d0c59719509220359220747 and https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1199/commits/f3878fda11737633ef9a501036de55153e80ffad.

**Comment by utpilla** [comment]
There is an unrelated test failing: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/5127460320/jobs/9224407051?pr=1199#step:7:213

**Pull request merged**